### PR TITLE
expression: handle max allowed packet for cast_as_binary

### PR DESCRIPTION
**Pull request by zz-jason**
### What problem does this PR solve? 

This PR fixes the incorrect result of the sql:
```sql
SELECT CAST( 'a' AS BINARY(429496729));
```

Result of TiDB:
```sql
TiDB(localhost:4000) > SELECT CAST( 'a' AS BINARY(429496729));
ERROR 2020 (HY000): Got packet bigger than 'max_allowed_packet' bytes
```

Result of MySQL:
```sql
MySQL(localhost:3306) > SELECT CAST( 'a' AS BINARY(429496729));
+---------------------------------+
| CAST( 'a' AS BINARY(429496729)) |
+---------------------------------+
| NULL                            |
+---------------------------------+
1 row in set, 1 warning (0.01 sec)

MySQL(localhost:3306) > show warnings;
+---------+------+-------------------------------------------------------------------------------------+
| Level   | Code | Message                                                                             |
+---------+------+-------------------------------------------------------------------------------------+
| Warning | 1301 | Result of cast_as_binary() was larger than max_allowed_packet (4194304) - truncated |
+---------+------+-------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### What is changed and how it works?

handle max allowed packet in the `cast_as_binary` function.

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
conflicts needs be resolved @zz-jason

File: ddl/session_pool.go [hunk 1]
```diff
@@ -44,6 +45,7 @@ func (sg *sessionPool) get() (sessionctx.Context, error) {
 	ctx := resource.(sessionctx.Context)
 	ctx.GetSessionVars().SetStatusFlag(mysql.ServerStatusAutocommit, true)
 	ctx.GetSessionVars().InRestrictedSQL = true
+	ctx.GetSessionVars().GetSystemVar(variable.MaxAllowedPacket)
```
Comment (zimulala) [inline ddl/session_pool.go]: Remove this line.

File: expression/integration_test.go [hunk 2]
```diff
@@ -936,6 +936,9 @@ func (s *testIntegrationSuite) TestStringBuiltin(c *C) {
 	result = tk.MustQuery("select a,b,concat_ws(',',a,b) from t")
 	result.Check(testkit.Rows("114.57011441 38.04620115 114.57011441,38.04620115",
 		"-38.04620119 38.04620115 -38.04620119,38.04620115"))
+
+	result = tk.MustQuery("SELECT CAST('a' AS BINARY(67108865));")
+	result.Check(testkit.Rows("<nil>"))
```
Comment (XuHuaiyu) [inline expression/integration_test.go]: check the result of `show warnings` here?

File: expression/builtin_cast.go [hunk 3]
```diff
@@ -1021,7 +1034,25 @@ func (b *builtinCastStringAsStringSig) evalString(row chunk.Row) (res string, is
 		return res, isNull, errors.Trace(err)
 	}
 	sc := b.ctx.GetSessionVars().StmtCtx
+
+	//	isTooLarge := uint64(len(res)) > b.maxAllowedPacket || (b.tp.Flen != types.UnspecifiedLength && uint64(b.tp.Flen) > b.maxAllowedPacket)
```
Comment (lamxTyler) [inline expression/builtin_cast.go]: Remove them?

File: expression/builtin_cast.go [hunk 4]
```diff
@@ -293,7 +299,12 @@ func (c *castAsStringFunctionClass) getFunction(ctx sessionctx.Context, args []E
 		sig = &builtinCastJSONAsStringSig{bf}
 		sig.setPbCode(tipb.ScalarFuncSig_CastJsonAsString)
 	case types.ETString:
-		sig = &builtinCastStringAsStringSig{bf}
+		valStr, _ := ctx.GetSessionVars().GetSystemVar(variable.MaxAllowedPacket)
+		maxAllowedPacket, err := strconv.ParseUint(valStr, 10, 64)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		sig = &builtinCastStringAsStringSig{bf, maxAllowedPacket}
```
Comment (zimulala) [inline expression/builtin_cast.go]: Could these be extracted into a function?

**Comment by zz-jason** [comment]
duplicated with https://github.com/pingcap/tidb/pull/8768, close it now.

**Pull request closed without merging**